README.md
## This is the Shell app for OnDemand 1.6. Future development has been moved to https://github.com/OSC/ondemand/tree/master/apps/shell

# OOD Shell

![GitHub Release](https://img.shields.io/github/release/osc/ood-shell.svg)
![GitHub License](https://img.shields.io/github/license/osc/ood-shell.svg)

This app is a Node.js app for Open OnDemand providing a web based terminal
using Chrome OS's hterm. It is meant to be run as the user (and on behalf of
the user) using the app. Thus, at an HPC center if I log into OnDemand using
the `ood` account, this app should run as `ood`.

## New Install

1.  Start in the **build directory** for all sys apps, clone and check out the
    latest version of the shell app (make sure the app directory's name is
    `shell`):

    ```sh
    scl enable rh-git29 -- git clone https://github.com/OSC/ood-shell.git shell
    cd shell
    scl enable rh-git29 -- git checkout tags/v1.4.2
    ```

2.  Install the app:

    ```sh
    scl enable rh-git29 rh-ruby24 rh-nodejs6 -- bin/setup
    ```

3.  Copy the built app directory to the deployment directory, and start the
    server. i.e.:

    ```sh
    sudo mkdir -p /var/www/ood/apps/sys
    sudo cp -r . /var/www/ood/apps/sys/shell
    ```

## Updating to a New Stable Version

1.  Navigate to the app's build directory and check out the latest version:

    ```sh
    cd shell # cd to build directory
    scl enable rh-git29 -- git fetch
    scl enable rh-git29 -- git checkout tags/v1.4.2
    ```

2.  Update the app:

    ```sh
    scl enable rh-git29 rh-ruby24 rh-nodejs6 -- bin/setup
    ```

3.  Copy the built app directory to the deployment directory:

    ```sh
    sudo rsync -rlptv --delete . /var/www/ood/apps/sys/shell
    ```

## Configuration

The app can be configured by editing the global environment file:

```
/etc/ood/config/apps/shell/env
```

An example environment file is:

```sh
# /etc/ood/config/apps/shell/env

# The default ssh host the user is logged into if the user doesn't specify a
# host in the url
DEFAULT_SSHHOST="localhost"
```

## Usage

Open a terminal to the default SSH host:

`http://localhost:3000/`

To specify the host:

`http://localhost:3000/ssh/<host>`

To specify a directory on the default host:

`http://localhost:3000/ssh/default/<dir>`

To specify a host and directory:

`http://localhost:3000/ssh/<host>/<dir>`

## Development

For development purposes the environment variables must be specified in the
local environment file:

```
.env.local
```

underneath the root directory of this app in your sandbox.

To mimic the production environment you may have to copy the production
environment variables down or set up a symbolic link:

```sh
# Copy production env vars
cp /etc/ood/config/apps/shell/env .env.local

# or setup a symlink
ln -s /etc/ood/config/apps/shell/env .env.local
```

Any changes made to the environment files require an app restart in order for
the changes to take effect:

```console
$ touch tmp/restart.txt
```

### Updating `hterm`

Clone Google's repository that includes some other things in addition to hterm:

```console
$ git clone https://chromium.googlesource.com/apps/libapps
```

Run the build script. It requires Python, specifically Python 3 for hterm 1.81 and newer. Here, it is run from the root directory of this new local repository (libapps):

```console
$ scl enable rh-python35 -- hterm/bin/mkdist.sh
```

There will be a file created in `hterm/dist/js` called `hterm_all.js`. Copy and rename this file to `public/javascripts/hterm_all_x.xx.js` in the Shell App repository, where x.xx represents the version number hterm (for cache busting), and change the reference in `views/index.hbs` to point to this new file.

#### Hacking `hterm`

So you've updated hterm and now something is broken.  Maybe only on one platform (*cough firefox*).  Here's a list of hacks/changes we've made to these files so when you're updating from one version to another you may have to add these 
changes. If you're lucky you may be able to cherry pick them, if not, hopefully we've made an issue where you can reference and you can at least see the commit. 

* 0cbc84e3d53386064e278a0495c940a217f4f18b - that fixed [issue 64](https://github.com/OSC/ood-shell/issues/64)

## Contributing

Bug reports and pull requests are welcome on GitHub at
https://github.com/OSC/ood-shell.

## License

The gem is available as open source under the terms of the [MIT
License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
## [1.5.1] - 2020-03-30
### Fixed
- downgrade node-pty dependency to version that is compatible with nodejs6

## [1.5.0] - 2020-03-30
### Fixed
- Add CSRF protection via CSRF token and Origin checking when creating shell app
  websocket connection [#444](https://github.com/OSC/ondemand/commit/1816de76fdf8bcec21d5f9619f5a3a09ff8db01d)
  backported to OnDemand 1.6 [#77](https://github.com/OSC/ood-shell/pull/77)
- Switch to yarn for dependency management
- Update dependencies

### Added
- Configurable SSH Wrapper for Shell app @baverhey [#406](https://github.com/OSC/ondemand/pull/406)
  backported to OnDemand 1.6 [#77](https://github.com/OSC/ood-shell/pull/77)

## [1.4.8] - 2019-10-09
### Changed
- revert previous fix and now omit `cwd` from spawn all the time. Since we only ssh now,
  changing $PWD before executing ssh does not matter

## [1.4.7] - 2019-10-08
### Fixed
- Omit `cwd` from spawn if it does not exist [#72](https://github.com/OSC/ood-shell/pull/72)

## [1.4.6] - 2019-06-25
### Fixed
- Fixed caching issue with `ood_shell.js`

## [1.4.5] - 2019-06-20
### Fixed
- Fixed bugs when running Shell in Firefox and MS Edge ([#62](https://github.com/OSC/ood-shell/issues/62), [#59](https://github.com/OSC/ood-shell/issues/59), and [#64](https://github.com/OSC/ood-shell/issues/64))

## [1.4.4] - 2019-06-17
### Changed
- Updated ws to address CVE

## [1.4.3] - 2019-05-09
### Changed
- Changed from pty.js to node-pty as it is still being maintained

## [1.4.2] - 2018-12-19
### Fixed
- Fixed cut and paste not working in Firefox [#48](https://github.com/OSC/ood-shell/issues/48)

## [1.4.1] - 2018-12-03
### Changed
- Updated hterm to 1.80 [#49](https://github.com/OSC/ood-shell/issues/49)

## [1.4.0] - 2018-09-11
### Changed
- Remove special case for "localhost". All shell sessions now use ssh.

## [1.3.1] - 2018-01-10
### Changed
- Remove and rebuild `node-modules` during every setup.
- Updated hterm from 1.73 to 1.77.
- Updated date in `LICENSE.md`.

## [1.3.0] - 2017-12-21
### Added
- Added support to read local environment variable file `.env.local`.
  [#43](https://github.com/OSC/ood-shell/pull/43)

### Changed
- Reads environment variables from global environment file
  `/etc/ood/config/apps/shell/env` in production mode.
  [#42](https://github.com/OSC/ood-shell/issues/42)

### Deprecated
- Deprecating old environment variable file `.env` located underneath the app.
  [#43](https://github.com/OSC/ood-shell/pull/43)

## [1.2.4] - 2017-10-20
### Changed
- Updated hterm from 1.61 to 1.73.
  [#41](https://github.com/OSC/ood-shell/issues/41)

## [1.2.3] - 2017-07-10
### Changed
- Changed the `CHANGELOG.md` formatting.

### Fixed
- Warn user if fail to establish websocket connection.
  [#38](https://github.com/OSC/ood-shell/issues/38)

## 1.2.2 - 2017-05-30
### Fixed
- Fix to handle multibyte UTF-8 URI decoding.
- Warn users if they try to close an active terminal.
- Warn users when their websocket connection is terminated.

[Unreleased]: https://github.com/OSC/ood-shell/compare/v1.5.1...HEAD
[1.5.1]: https://github.com/OSC/ood-shell/compare/v1.5.0...v1.5.1
[1.5.0]: https://github.com/OSC/ood-shell/compare/v1.4.8...v1.5.0
[1.4.8]: https://github.com/OSC/ood-shell/compare/v1.4.7...v1.4.8
[1.4.7]: https://github.com/OSC/ood-shell/compare/v1.4.6...v1.4.7
[1.4.6]: https://github.com/OSC/ood-shell/compare/v1.4.5...v1.4.6
[1.4.5]: https://github.com/OSC/ood-shell/compare/v1.4.4...v1.4.5
[1.4.4]: https://github.com/OSC/ood-shell/compare/v1.4.3...v1.4.4
[1.4.3]: https://github.com/OSC/ood-shell/compare/v1.4.2...v1.4.3
[1.4.2]: https://github.com/OSC/ood-shell/compare/v1.4.1...v1.4.2
[1.4.1]: https://github.com/OSC/ood-shell/compare/v1.4.0...v1.4.1
[1.4.0]: https://github.com/OSC/ood-shell/compare/v1.3.1...v1.4.0
[1.3.1]: https://github.com/OSC/ood-shell/compare/v1.3.0...v1.3.1
[1.3.0]: https://github.com/OSC/ood-shell/compare/v1.2.4...v1.3.0
[1.2.4]: https://github.com/OSC/ood-shell/compare/v1.2.3...v1.2.4
[1.2.3]: https://github.com/OSC/ood-shell/compare/v1.2.2...v1.2.3


public/javascripts/ood_shell.js

// Object that defines a terminal element
function OodShell(element, url, prefs) {
  this.element = element;
  this.url     = url;
  this.prefs   = prefs || {};
  this.socket  = null;
  this.term    = null;
}

/**
 * Create the terminal. The constructor for the OodShell class is mostly just a data object. This function
 * does the actual work of initializing the hterm.Terminal, the websocket (through a callback *after* the 
 * terminal is ready.) and setting preferences, etc.
 */
OodShell.prototype.createTerminal = function () {
  // Set backing store that hterm uses to read/write preferences
  hterm.defaultStorage = new lib.Storage.Memory();
  
  // Create an instance of hterm.Terminal
  this.term = new hterm.Terminal();
    
  var that = this;
    
  // Handler that fires when terminal is initialized and ready for use
  this.term.onTerminalReady = function () {
    // Create a new terminal IO object and give it the foreground.
    // (The default IO object just prints warning messages about unhandled
    // things to the JS console.)
    var io = this.io.push();
  
    // Set up event handlers for io
    io.onVTKeystroke    = that.onVTKeystroke.bind(that);
    io.sendString       = that.sendString.bind(that);
    io.onTerminalResize = that.onTerminalResize.bind(that);
    
    that.socket = new WebSocket(that.url);  // WS isn't made or bound until terminal is ready
    that.socket.onopen    = that.useCachedSize.bind(that); //we've been caching the term size while connecting
    that.socket.onmessage = that.getMessage.bind(that);
    that.socket.onclose   = that.closeTerminal.bind(that);
    
    // Capture all keyboard input
    this.installKeyboard();
  };
    
  // Patch cursor setting
  this.term.options_.cursorVisible = true;
    
  // Connect terminal to sacrificial DOM node
  this.term.decorate(this.element);
    
  // Set preferences for terminal
  for (var k in this.prefs) {
    this.term.prefs_.set(k, this.prefs[k]);
  }
    
  // Warn user if he/she unloads page
  window.onbeforeunload = function() {
    return 'Leaving this page will terminate your terminal session.';
  };
};

OodShell.prototype.getMessage = function (ev) {
  this.term.io.print(ev.data);
}

OodShell.prototype.closeTerminal = function (ev) {
  var errorDiv;

  // Do not need to warn user if he/she unloads page
  window.onbeforeunload = null;

  // Inform user they lost connection
  if ( this.term === null ) {
    errorDiv = document.createElement('div');
    errorDiv.className = 'error';
    errorDiv.innerHTML = 'Failed to establish a websocket connection. Be sure you are using a browser that supports websocket connections.';
    this.element.appendChild(errorDiv);
  } else {
    this.term.io.print('\r\nYour connection to the remote server has been terminated.');
  }
}

OodShell.prototype.onVTKeystroke = function (str) {
  // Do something useful with str here.
  // For example, Secure Shell forwards the string onto the NaCl plugin.
  this.socket.send(JSON.stringify({
    input: str
  }));
};

OodShell.prototype.sendString = function (str) {
  // Just like a keystroke, except str was generated by the
  // terminal itself.
  // Most likely you'll do the same this as onVTKeystroke.
  this.onVTKeystroke(str)
};

OodShell.prototype.onTerminalResize = function (columns, rows) {
  // React to size changes here.
  // Secure Shell pokes at NaCl, which eventually results in
  // some ioctls on the host.
  
  // iff the socket is ready. race condition in FF where we try to resize before the socket is open
  if(this.socket.readyState == 1){ 
    this.socket.send(JSON.stringify({
      resize: {
        cols: columns,
        rows: rows
      }
    }));
  }else{
      this.cachedCols = columns;
      this.cachedRows = rows;
  }

};

OodShell.prototype.useCachedSize = function() {
  if(this.cachedCols && this.cachedRows) {
    this.onTerminalResize(this.cachedCols, this.cachedRows);
  }
};


app.js
var fs        = require('fs');
var http      = require('http');
var path      = require('path');
var WebSocket = require('ws');
var express   = require('express');
var pty       = require('node-pty');
var hbs       = require('hbs');
var dotenv    = require('dotenv');
var Tokens    = require('csrf');
var url       = require('url');
var port      = 3000;
var URLSearchParams = require('whatwg-url').URLSearchParams;


// Read in environment variables
dotenv.config({path: '.env.local'});
if (process.env.NODE_ENV === 'production') {
  dotenv.config({path: '/etc/ood/config/apps/shell/env'});
}

// Keep app backwards compatible
if (fs.existsSync('.env')) {
  console.warn('[DEPRECATION] The file \'.env\' is being deprecated. Please move this file to \'/etc/ood/config/apps/shell/env\'.');
  dotenv.config({path: '.env'});
}

const tokens = new Tokens({});
const secret = tokens.secretSync();

// Create all your routes
var router = express.Router();
router.get('/', function (req, res) {
  res.redirect(req.baseUrl + '/ssh');
});

router.get('/ssh*', function (req, res) {
  res.render('index',
    {
      baseURI: req.baseUrl,
      csrfToken: tokens.create(secret),
    });
});

router.use(express.static(path.join(__dirname, 'public')));

// Setup app
var app = express();

// Setup template engine
app.set('view engine', 'hbs');
app.set('views', path.join(__dirname, 'views'));

// Mount the routes at the base URI
app.use(process.env.PASSENGER_BASE_URI || '/', router);

// Setup websocket server
var server = new http.createServer(app);
var wss = new WebSocket.Server({ noServer: true });

wss.on('connection', function connection (ws, req) {
  var match;
  var host = process.env.DEFAULT_SSHHOST || 'localhost';
  var cmd = process.env.OOD_SSH_WRAPPER || 'ssh';
  var dir;
  var term;
  var args;
  var host_path_rx = '/ssh/([^\\/\\?]+)([^\\?]+)?(\\?.*)?$';

  console.log('Connection established');

  // Determine host and dir from request URL
  if (match = req.url.match(process.env.PASSENGER_BASE_URI + host_path_rx)) {
    if (match[1] !== 'default') host = match[1];
    if (match[2]) dir = decodeURIComponent(match[2]);
  }

  args = dir ? [host, '-t', 'cd \'' + dir.replace(/\'/g, "'\\''") + '\' ; exec ${SHELL} -l'] : [host];

  term = pty.spawn(cmd, args, {
    name: 'xterm-256color',
    cols: 80,
    rows: 30
  });

  console.log('Opened terminal: ' + term.pid);

  term.on('data', function (data) {
    ws.send(data, function (error) {
      if (error) console.log('Send error: ' + error.message);
    });
  });

  term.on('error', function (error) {
    ws.close();
  });

  term.on('close', function () {
    ws.close();
  });

  ws.on('message', function (msg) {
    msg = JSON.parse(msg);
    if (msg.input)  term.write(msg.input);
    if (msg.resize) term.resize(parseInt(msg.resize.cols), parseInt(msg.resize.rows));
  });

  ws.on('close', function () {
    term.end();
    console.log('Closed terminal: ' + term.pid);
  });
});

function custom_server_origin(default_value = null){
  var custom_origin = null;

  if(process.env.OOD_SHELL_ORIGIN_CHECK) {
    // if ENV is set, do not use default!
    if(process.env.OOD_SHELL_ORIGIN_CHECK.startsWith('http')){
      custom_origin = process.env.OOD_SHELL_ORIGIN_CHECK;
    }
  }
  else {
    custom_origin = default_value;
  }

  return custom_origin;
}

function default_server_origin(headers){
  var origin = null;

  if (headers['x-forwarded-proto'] && headers['x-forwarded-host']){
    origin = headers['x-forwarded-proto'] + "://" + headers['x-forwarded-host']
  }

  return origin;
}

server.on('upgrade', function upgrade(request, socket, head) {
  var requestToken = new URLSearchParams(url.parse(request.url).search).get('csrf'),
      client_origin = request.headers['origin'],
      server_origin = custom_server_origin(default_server_origin(request.headers));

  if (client_origin &&
      client_origin.startsWith('http') &&
      server_origin && client_origin !== server_origin
  ) {
    socket.write([
      'HTTP/1.1 401 Unauthorized',
      'Content-Type: text/html; charset=UTF-8',
      'Content-Encoding: UTF-8',
      'Connection: close',
      'X-OOD-Failure-Reason: invalid origin',
    ].join('\r\n') + '\r\n\r\n');

    socket.destroy();
  }
  else if (!tokens.verify(secret, requestToken)) {
    socket.write([
      'HTTP/1.1 401 Unauthorized',
      'Content-Type: text/html; charset=UTF-8',
      'Content-Encoding: UTF-8',
      'Connection: close',
      'X-OOD-Failure-Reason: bad csrf token',
    ].join('\r\n') + '\r\n\r\n');

    socket.destroy();

  }else{
    wss.handleUpgrade(request, socket, head, function done(ws) {
      wss.emit('connection', ws, request);
    });
  }
});

server.listen(port, function () {
  console.log('Listening on ' + port);
});
